skglm/solvers/prox_newton.py

```diff
@@ -0,0 +1,338 @@
+            p_obj = datafit.value(y, w, Xw) + penalty.value(w)
+            print(
+                f"Iteration {t+1}: {p_obj:.10f}, "
+                f"stopping crit: {stop_crit:.2e}"
+            )
+
+        if stop_crit <= tol:
+            if verbose:
+                print(f"Stopping criterion max violation: {stop_crit:.2e}")
+            break
+
+        # build working set
+        gsupp_size = penalty.generalized_support(w).sum()
+        ws_size = max(min(p0, n_features),
+                      min(n_features, 2 * gsupp_size))
+        # similar to np.argsort()[-ws_size:] but without sorting
+        ws = np.argpartition(opt, -ws_size)[-ws_size:]
+
+        grad_ws = grad[ws]
+        tol_in = EPS_TOL * stop_crit
+
+        for epoch in range(max_epochs):
+
+            # find descent direction
+            if is_sparse:
+                delta_w_ws, X_delta_w_ws = _descent_direction_s(
+                    *X_bundles, y, w, Xw, grad_ws, datafit,
+                    penalty, ws, tol=EPS_TOL*tol_in)
+            else:
+                delta_w_ws, X_delta_w_ws = _descent_direction(
+                    X, y, w, Xw, grad_ws, datafit, penalty, ws, tol=EPS_TOL*tol_in)
+
+            # backtracking line search with inplace update of w, Xw
+            if is_sparse:
+                grad_ws[:] = _backtrack_line_search_s(*X_bundles, y, w, Xw,
+                                                      datafit, penalty, delta_w_ws,
+                                                      X_delta_w_ws, ws)
+            else:
+                grad_ws[:] = _backtrack_line_search(X, y, w, Xw, datafit, penalty,
+                                                    delta_w_ws, X_delta_w_ws, ws)
+
+            # check convergence
+            opt_in = penalty.subdiff_distance(w, grad_ws, ws)
+            stop_crit_in = np.max(opt_in)
+
+            if max(verbose-1, 0):
+                p_obj = datafit.value(y, w, Xw) + penalty.value(w)
+                print(
+                    f"Epoch {epoch+1}: {p_obj:.10f}, "
+                    f"stopping crit in: {stop_crit_in:.2e}"
+                )
+
+            if stop_crit_in <= tol_in:
+                if max(verbose-1, 0):
+                    print("Early exit")
+                break
+
+        p_obj = datafit.value(y, w, Xw) + penalty.value(w)
+        p_objs_out.append(p_obj)
+    return w, p_objs_out, stop_crit
+
+
+@njit
+def _descent_direction(X, y, w_epoch, Xw_epoch, grad_ws, datafit,
+                       penalty, ws, tol):
+    # Given:
+    #   1) b = \nabla F(X w_epoch)
+    #   2) D = \nabla^2 F(X w_epoch)  <------>  raw_hess
+    # Minimize quadratic approximation for delta_w = w - w_epoch:
+    #  b.T @ X @ delta_w + \
+    #  1/2 * delta_w.T @ (X.T @ D @ X) @ delta_w + penalty(w)
+    raw_hess = datafit.raw_hessian(y, Xw_epoch)
+
+    lipschitz = np.zeros(len(ws))
+    for idx, j in enumerate(ws):
+        lipschitz[idx] = raw_hess @ X[:, j] ** 2
+
+    cached_grads = np.zeros(len(ws))
```

Comment:
do you know why we call this cached_grads ?

Reply:
cached_grads are the grads of the quadratic approximation, and we use them in 
- soft thresh update
- check convergence when solving the quadratic approximation

the cached prefix refers to the fact that we are computing them only once: When performing the soft thresh updates.

@mathurinm, do you have a better name?

Reply:
`old_grads` ?

Reply:
``quadratic_grads``?

Reply:
you need to insist on the fact that they are not up to date gradients